Add more infrastructure for timing

mwarusz opened the pull request:
<!--
Please add a description of what is accomplished in the PR here at the top:
-->
Adds a new module for timing under `infra/`. This module build upon the Pacer library, but extends it
with support for specifying timing levels and automatically adding Kokkos fences. It also includes more advanced
timer options that are useful in special circumstances. Finally, it adds integration with NVTX ranges on NVIDIA GPUs for profiling with vendor-specific tools.

To show how all of this can be used, new timers have been added in the `AuxiliaryState` and `Tendencies` modules.

The existing timers in `Decomp` still need to be converted to the new API, but I wanted to gather feedback before making changes there. 

For details on what was added and how to use it see the user and developer docs:
- User:  https://portal.nersc.gov/project/e3sm/mwarusz/timing/html/userGuide/Timing.html
- Developer: https://portal.nersc.gov/project/e3sm/mwarusz/timing/html/devGuide/Timing.html


<!--
Below are a few things we ask you or your reviewers to kindly check. 
***Remove checks that are not relevant by deleting the line(s) below.***
-->
Checklist
* [x] Documentation:
  * [x] User's Guide has been updated
  * [x] Developer's Guide has been updated
  * [x] Documentation has been [built locally](https://e3sm-project.github.io/Omega/omega/develop/devGuide/BuildDocs.html) and changes look as expected
* [x] Testing
  * [x] Unit tests have passed. Please provide a relevant CDash build entry for verification.



<!--
Please note any issues this fixes using closing keywords: https://help.github.com/articles/closing-issues-using-keywords
-->

mwarusz committed:
Fix ERROR_REQUIRE and ERROR_ASSERT macros

mwarusz commented:
> This is not an issue, but I’m curious if the timing "Level" is fixed as currently defined in the source code of this PR. For example, routines in the Halo module have a timing level of 4. Will it always remain Level 4, even if new Omega code is introduced that uses Halo functions?

@grnydawn 
We need to decide what timing level we want by default and come up with some guidelines on what to include at different timing levels. What I have in this PR is not mean to be final.

philipwjones reviewed (changes requested):
I like the new features, but we probably should have had a design doc/discussion to decide some of the issues below before it got to a PR.  So a few comments/changes:

- at some point, I think some of these new features are better incorporated in the underlying Pacer routines. I worry that this is creating a 3-deep hierarchy in what really needs to be a very lightweight call (Timing calls Pacer which calls gptl/other low-level profiler). We could migrate them later, but should have that discussion with @sarats 
- we can't use 'MPI_COMM_WORLD' or this will deadlock the coupled system when atm and ocn are run asynchronously (as is typical). So we need to either retrieve the ocn communicator from the default MachEnv or (maybe more desirable?) pass in either a MachEnv or communicator as an argument during init. And the timers will need to retain that communicator somewhere.
- I think we're going to need the MPI barrier argument to specify where a barrier might be needed and then either a global flag (like the fence) or the timer level to determine whether to actually call the barrier.  We often like timers around the comm routines to be part of our standard profile we track for every run (see timer level discussion) but do not want the barrier activated unless we really need the specific comm time without wait time included.
- I will start a separate thread on timer levels since that requires some discussion

sarats commented:
@mwarusz  As Phil noted, it would have been best to have discussed these features before doing the work.

In fact, many of the features (timer depth, MPI barriers etc.) are already present in GPTL. That was the primary reason I kept Pacer as a thin layer over GPTL. 
perf_mod has a number of options to control the timing infrastructure with sane defaults as well as a way to override those by specifying in drv_in (prof_inparm section)
https://github.com/E3SM-Project/E3SM/blob/master/share/timing/perf_mod.F90#L87

I have long debated whether it makes sense for us to put NVTX and ROCTx directly into Pacer or use something like PerfStubs (https://github.com/UO-OACISS/perfstubs) which brings in external performance tool integration (Tau/Apex as well as GPU profiling).

Furthermore, the new features that are desired are certainly better placed in the Pacer layer. Currently, Omega is the only component using Pacer, so we don't have any version freeze or backwards compatibility to worry about. I'm not clear about the need for vector vs. hashmap for tracking open timers. I'm not gatekeeping Pacer and would be glad to have you contribute/update the design and data structures to suit Omega's needs.

mwarusz committed:
Use updated Pacer

mwarusz committed:
Add a getter for Halo MPI comm

mwarusz committed:
More timers

mwarusz committed:
Add CMake option to add ranges for external profilers

mwarusz committed:
Add PrintAllRanks option in timing config

mwarusz commented:
@philipwjones 

I have updated this PR and moved most of the new functionality to the Pacer library. This PR should probably be split into a PR to E3SM with the Pacer changes and a PR to Omega with the Omega changes. However, before I do that and update the docs, I would appreciate it if you could have a look at the current state of this PR and confirm that it addresses your comments.

Here is a summary of the changes:
- Modified Pacer to support timing levels, automatic Kokkos fences, NVTX ranges, adding parent prefixes, and 
  disabling timing in a region.
- Kokkos integration and NVTX ranges in Pacer are optional (they are guarded by ifdefs).
- The options to disable timing or add prefixes are no longer arguments to the start and stop functions. Instead, they are controlled by the free functions pairs `Pacer::disableTiming()`/`Pacer::enableTiming()` and `Pacer::addParentPrefix()`/`Pacer::removeParentPrefix()`.
- A new Pacer function `timingBarrier(Name, Level, Comm)` was added. It optionally calls and times an MPI barrier on the provided communicator. This can be enabled by the `Pacer::enableTimingBarriers()` function.
- Other than adding timers, most of the changes in Omega are in `OceanInit.cpp`, where the reading and setting of Pacer options now happens.
- The timing config options are now: `Level`, `AutoFence`, `TimingBarriers`, and `PrintAllRanks`.

grnydawn reviewed (approved):
Tests on Frontier and Perlmutter-GPU have passed. Approved.

mwarusz committed:
Add Pacer::finalize calls

grnydawn commented:
@mwarusz , I ran ctest of `mwarusz:omega/timing-infra` branch on Frontier and got core dumps from the following tests:
```
	  8 - HALO_TEST (Failed)
	 10 - HORZMESH_TEST (Failed)
	 12 - HORZOPERATORS_SPHERE_TEST (Failed)
	 13 - AUXVARS_PLANE_TEST (Failed)
	 15 - AUXSTATE_TEST (Failed)
	 20 - TEND_PLANE_TEST (Failed)
	 23 - TENDENCIES_TEST (Failed)
	 27 - TIMESTEPPER_TEST (Failed)
```
GDB backtrace shows that the issue is caused from:
```
#5  0x00000000007706ca in Pacer::addParentPrefix() ()
```

Could you look at this issue? FYI, `craycray-mphipcc` compiler is used to build HIP version of Omega.

mwarusz committed:
Add timing docs

mwarusz committed:
Comment out buggy Pacer functions

mwarusz commented:
@grnydawn Addressing your earlier comments 
>     1. I noticed that the level of these two timers is set to zero. I just want to confirm that this is what we decided:
>        ```
>        base/Decomp.cpp:   bool TimerFlag = Pacer::start("Decomp init", 0);
>        base/Decomp.cpp:   TimerFlag = Pacer::stop("Decomp init", 0) && TimerFlag;
>        infra/Config.cpp:  Pacer::start("ConfigReadAll", 0);
>        infra/Config.cpp:  Pacer::stop("ConfigReadAll", 0);
>        ```

I think level 0 for "Decomp init" makes sense. "ConfigReadAll" is tricky, because at that point we haven't read the timing config yet. So level 0 is the only option for this timer to show up.   

>     2. I found that none of the Omega tests call `Pacer::finalize();`. It may not cause any issues, but I think it would be good to add this at the end of each test.

I added calls to `Pacer::finalize()` in every test.

>     3. I also noticed that none of the Omega tests call the `Pacer::print` function. I think it would be useful to add this function call in at least some of the Omega tests so that we can easily get timing information from those runs.

Possibly, but let's do it later.

> Could you look at this issue? FYI, craycray-mphipcc compiler is used to build HIP version of Omega.

With regards to the test failures on Frontier, I have tracked them down to an edge case in `Pacer::addParentPrefix()` which I forgot to handle. Since this Pacer function is used only in the halo exchange I have decided to comment it out for now to get `develop` to a working state quickly. Fixing this edge case is simple, but it will require another E3SM PR and  upstream merge.

mwarusz commented:
@philipwjones @grnydawn  

I have rebased this PR and added documentation. There is a bug in `Pacer::addParentPrefix()` and `Pacer:removeParentPrefix()` that needs to be fixed in E3SM, so I commented out calls to these functions. I tested on Frontier and Perlmutter both cpu and gpu and everything is passing. I believe this PR is ready to be merged.

grnydawn commented:
@mwarusz , Thanks for the quick updates. The changes have passed CTests on Frontier with several compilers.

philipwjones commented:
Thanks @mwarusz and @grnydawn  I'll go ahead and merge
